src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long long unsigned int' [-fpermissive]

### The problem

src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long long unsigned int' [-fpermissive] 

### Which version of ESPHome has the issue?

2025.7.1

### What type of installation are you using?

Docker

### What platform are you using?

ESP8266

### Component causing the issue

_No response_

### YAML Config

```yaml
substitutions:
  device_name: rf-bridge-01
  friendly_name: RF Bridge 01
  comment: Sonoff RF Bridge | IP DHCP

esphome:
  name: ${device_name}
  friendly_name: ${friendly_name}

esp8266:
  board: esp01_1m

logger:

api:
  encryption:
    key: "removed"

ota:
  - platform: esphome
    password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  ap:
    ssid: "Rf-Bridge 01 Hotspot"
    password: !secret ap_password

web_server:
  port: 80
  version: 3

time:
  - platform: homeassistant
    id: homeassistant_time

captive_portal:

packages:
  text_sensor: !include packages/text_sensor.yaml
  button: !include packages/button.yaml
  sensor: !include packages/sensor.yaml

################################################################################

rf_bridge:
  on_code_received:
    then:
      - homeassistant.event:
          event: esphome.rf_code_received01
          data:
            sync: !lambda 'return format_hex(data.sync);'
            low: !lambda 'return format_hex(data.low);'
            high: !lambda 'return format_hex(data.high);'
            code: !lambda 'return format_hex(data.code);'
            
      - text_sensor.template.publish:
          id: received_code01
          state: !lambda 'char buffer [16];return itoa(data.code,buffer,16);'

      - delay: 1s

      - text_sensor.template.publish:
          id: received_code01
          state: "0"

  on_advanced_code_received:
    then:
      - homeassistant.event:
          event: esphome.rf_advanced_code_received01
          data:
            length: !lambda 'char buffer [10];return itoa(data.length,buffer,16);'
            protocol: !lambda 'char buffer [10];return itoa(data.protocol,buffer,16);'
            code: !lambda 'return data.code;'        

uart:
  tx_pin: 1
  rx_pin: 3
  baud_rate: 19200

text_sensor:

  - platform: template
    id: received_code01
    icon: "mdi:lightbulb"
    name: "code automation"
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2025.7.1
INFO Reading configuration /config/RF Bridge 01.yaml...
INFO Detected timezone 'Europe/Warsaw'
INFO Generating C++ source...
INFO Compiling app...
Processing rf-bridge-01 (board: esp01_1m; framework: arduino; platform: platformio/espressif8266@4.2.1)
Library Manager: Installing ESP32Async/ESPAsyncTCP @ 2.0.0
INFO Installing ESP32Async/ESPAsyncTCP @ 2.0.0
Unpacking [####################################] 100%
Library Manager: ESPAsyncTCP@2.0.0 has been installed!
INFO ESPAsyncTCP@2.0.0 has been installed!
Library Manager: Installing ESP32Async/ESPAsyncWebServer @ 3.7.10
INFO Installing ESP32Async/ESPAsyncWebServer @ 3.7.10
Unpacking [####################################] 100%
Library Manager: ESPAsyncWebServer@3.7.10 has been installed!
INFO ESPAsyncWebServer@3.7.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/noise-c @ 0.1.10
INFO Installing esphome/noise-c @ 0.1.10
Unpacking [####################################] 100%
Library Manager: noise-c@0.1.10 has been installed!
INFO noise-c@0.1.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10020.7
INFO Installing esphome/libsodium @ 1.10020.7
Unpacking [####################################] 100%
Library Manager: libsodium@1.10020.7 has been installed!
INFO libsodium@1.10020.7 has been installed!
Library Manager: Installing bblanchon/ArduinoJson @ 7.4.2
INFO Installing bblanchon/ArduinoJson @ 7.4.2
Unpacking [####################################] 100%
Library Manager: ArduinoJson@7.4.2 has been installed!
INFO ArduinoJson@7.4.2 has been installed!
HARDWARE: ESP8266 80MHz, 80KB RAM, 1MB Flash
Dependency Graph
|-- ESPAsyncTCP @ 2.0.0
|-- ESP8266WiFi @ 1.0
|-- ESPAsyncWebServer @ 3.7.10
|-- DNSServer @ 1.1.1
|-- ESP8266mDNS @ 1.2
|-- noise-c @ 0.1.10
|-- ArduinoJson @ 7.4.2
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/button/button.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/esp8266/core.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/esp8266/gpio.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/esp8266/helpers.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/esp8266/preferences.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/homeassistant/time/homeassistant_time.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/logger/logger_esp8266.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/restart/button/restart_button.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/rf_bridge/rf_bridge.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/socket/lwip_raw_tcp_impl.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/template/text_sensor/template_text_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/text_sensor/filter.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/text_sensor/text_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/time/automation.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/time/real_time_clock.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uart/uart.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uart/uart_component.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uart/uart_component_esp8266.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uart/uart_debugger.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uptime/sensor/uptime_seconds_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/uptime/sensor/uptime_timestamp_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/version/version_text_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/web_server/list_entities.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/web_server/ota/ota_web_server.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/web_server/web_server.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/web_server/web_server_v1.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/wifi_info/wifi_info_text_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/components/wifi_signal/wifi_signal_sensor.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/application.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/color.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/component.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/controller.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/log.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/string_ref.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/time.cpp.o
Compiling .pioenvs/rf-bridge-01/src/esphome/core/util.cpp.o
Compiling .pioenvs/rf-bridge-01/src/main.cpp.o
Generating LD script .pioenvs/rf-bridge-01/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/rf-bridge-01/libebd/ESPAsyncTCP/AsyncPrinter.cpp.o
Compiling .pioenvs/rf-bridge-01/libebd/ESPAsyncTCP/ESPAsyncTCP.cpp.o
Compiling .pioenvs/rf-bridge-01/libebd/ESPAsyncTCP/ESPAsyncTCPbuffer.cpp.o
Compiling .pioenvs/rf-bridge-01/libebd/ESPAsyncTCP/SyncClient.cpp.o
Compiling .pioenvs/rf-bridge-01/libebd/ESPAsyncTCP/tcp_axtls.c.o
Compiling .pioenvs/rf-bridge-01/libfb1/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/rf-bridge-01/libfb1/ESP8266WiFi/CertStoreBearSSL.cpp.o
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h: In instantiation of 'static std::string esphome::api::TemplatableStringValue::value_to_string(T&&) [with T = char*; X = {esphome::rf_bridge::RFBridgeAdvancedData}; std::string = std::__cxx11::basic_string]':
src/esphome/components/api/homeassistant_service.h:31:96: required from 'esphome::api::TemplatableStringValue::TemplatableStringValue(F) [with F = setup()::<lambda(esphome::rf_bridge::RFBridgeAdvancedData)>; typename std::enable_if<std::is_invocable<F, X ...>::value, int>::type = 0; X = {esphome::rf_bridge::RFBridgeAdvancedData}]'
src/esphome/components/api/homeassistant_service.h:36:108: required from 'esphome::api::TemplatableKeyValuePair::TemplatableKeyValuePair(std::string, T) [with T = setup()::<lambda(esphome::rf_bridge::RFBridgeAdvancedData)>; Ts = {esphome::rf_bridge::RFBridgeAdvancedData}; std::string = std::__cxx11::basic_string]'
src/esphome/components/api/homeassistant_service.h:48:27: required from 'void esphome::api::HomeAssistantServiceCallAction::add_data(std::string, T) [with T = setup()::<lambda(esphome::rf_bridge::RFBridgeAdvancedData)>; Ts = {esphome::rf_bridge::RFBridgeAdvancedData}; std::string = std::__cxx11::basic_string]'
/config/RF Bridge 01.yaml:78:4: required from here
src/esphome/components/api/homeassistant_service.h:16:86: error: no matching function for call to 'to_string(char*)'
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6649:3: note: candidate: 'std::string std::__cxx11::to_string(long long unsigned int)' (near match)
6649 | to_string(unsigned long long __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6649:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long long unsigned int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6637:3: note: candidate: 'std::string std::__cxx11::to_string(long long int)' (near match)
6637 | to_string(long long __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6637:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long long int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6629:3: note: candidate: 'std::string std::__cxx11::to_string(long unsigned int)' (near match)
6629 | to_string(unsigned long __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6629:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long unsigned int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6618:3: note: candidate: 'std::string std::__cxx11::to_string(long int)' (near match)
6618 | to_string(long __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6618:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'long int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6610:3: note: candidate: 'std::string std::__cxx11::to_string(unsigned int)' (near match)
6610 | to_string(unsigned __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6610:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'unsigned int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
In file included from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/string:55,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/locale_classes.h:40,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ios_base.h:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/streambuf:41,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/streambuf_iterator.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/iterator:66,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algobase.h:36,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/ranges_algo.h:35,
from /config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/algorithm:64,
from /config/.esphome/platformio/packages/framework-arduinoespressif8266/cores/esp8266/Arduino.h:229,
from src/esphome/core/macros.h:7,
from src/esphome.h:2,
from src/main.cpp:3:
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6599:3: note: candidate: 'std::string std::__cxx11::to_string(int)' (near match)
6599 | to_string(int __val)
| ^~~~~~~~~
/config/.esphome/platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/10.3.0/bits/basic_string.h:6599:3: note: conversion of argument 1 would be ill-formed:
In file included from src/esphome.h:10,
from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h:16:86: error: invalid conversion from 'char*' to 'int' [-fpermissive]
16 | template static std::string value_to_string(T &&val) { return to_string(std::forward(val)); }
| ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
| |
| char*
Compiling .pioenvs/rf-bridge-01/libfb1/ESP8266WiFi/ESP8266WiFi.cpp.o
Archiving .pioenvs/rf-bridge-01/libebd/libESPAsyncTCP.a
Compiling .pioenvs/rf-bridge-01/libfb1/ESP8266WiFi/ESP8266WiFiAP.cpp.o
*** [.pioenvs/rf-bridge-01/src/main.cpp.o] Error 1
========================= [FAILED] Took 237.27 seconds =========================
```

### Additional information

_No response_

--- a/esphome/components/api/homeassistant_service.h
+++ b/esphome/components/api/homeassistant_service.h
@@ -17,7 +17,7 @@ template<typename... X> class TemplatableStringValue : public TemplatableValue<s
 
   // Overloads for string types - needed because std::to_string doesn't support them
   static std::string value_to_string(char *val) {
-    return std::string(val);
+    return val ? std::string(val) : std::string();
   }  // For lambdas returning char* (e.g., itoa)
   static std::string value_to_string(const char *val) { return std::string(val); }  // For lambdas returning .c_str()
   static std::string value_to_string(const std::string &val) { return val; }

--- a/esphome/components/api/homeassistant_service.h
+++ b/esphome/components/api/homeassistant_service.h
@@ -16,6 +16,9 @@ template<typename... X> class TemplatableStringValue : public TemplatableValue<s
   template<typename T> static std::string value_to_string(T &&val) { return to_string(std::forward<T>(val)); }
 
   // Overloads for string types - needed because std::to_string doesn't support them
+  static std::string value_to_string(char *val) {
+    return std::string(val);
+  }  // For lambdas returning char* (e.g., itoa)
   static std::string value_to_string(const char *val) { return std::string(val); }  // For lambdas returning .c_str()
   static std::string value_to_string(const std::string &val) { return val; }
   static std::string value_to_string(std::string &&val) { return std::move(val); }

--- a/tests/integration/fixtures/api_string_lambda.yaml
+++ b/tests/integration/fixtures/api_string_lambda.yaml
@@ -60,5 +60,28 @@ api:
             data:
               value: !lambda 'return input_float;'
 
+    # Service that tests char* lambda functionality (e.g., from itoa or sprintf)
+    - action: test_char_ptr_lambda
+      variables:
+        input_number: int
+        input_string: string
+      then:
+        # Log the input to verify service was called
+        - logger.log:
+            format: "Service called with number for char* test: %d"
+            args: [input_number]
+
+        # Test that char* lambdas work correctly
+        # This would fail in issue #9628 with "invalid conversion from 'char*' to 'long long unsigned int'"
+        - homeassistant.event:
+            event: esphome.test_char_ptr_lambda
+            data:
+              # Test snprintf returning char*
+              decimal_value: !lambda 'static char buffer[20]; snprintf(buffer, sizeof(buffer), "%d", input_number); return buffer;'
+              # Test strdup returning char* (dynamically allocated)
+              string_copy: !lambda 'return strdup(input_string.c_str());'
+              # Test string literal (const char*)
+              literal: !lambda 'return "test literal";'
+
 logger:
   level: DEBUG

--- a/tests/integration/test_api_string_lambda.py
+++ b/tests/integration/test_api_string_lambda.py
@@ -19,15 +19,17 @@ async def test_api_string_lambda(
     """Test TemplatableStringValue works with lambdas that return different types."""
     loop = asyncio.get_running_loop()
 
-    # Track log messages for all three service calls
+    # Track log messages for all four service calls
     string_called_future = loop.create_future()
     int_called_future = loop.create_future()
     float_called_future = loop.create_future()
+    char_ptr_called_future = loop.create_future()
 
     # Patterns to match in logs - confirms the lambdas compiled and executed
     string_pattern = re.compile(r"Service called with string: STRING_FROM_LAMBDA")
     int_pattern = re.compile(r"Service called with int: 42")
     float_pattern = re.compile(r"Service called with float: 3\.14")
+    char_ptr_pattern = re.compile(r"Service called with number for char\* test: 123")
 
     def check_output(line: str) -> None:
         """Check log output for expected messages."""
@@ -37,6 +39,8 @@ async def test_api_string_lambda(
             int_called_future.set_result(True)
         if not float_called_future.done() and float_pattern.search(line):
             float_called_future.set_result(True)
+        if not char_ptr_called_future.done() and char_ptr_pattern.search(line):
+            char_ptr_called_future.set_result(True)
 
     # Run with log monitoring
     async with (
@@ -65,17 +69,28 @@ async def test_api_string_lambda(
         )
         assert float_service is not None, "test_float_lambda service not found"
 
-        # Execute all three services to test different lambda return types
+        char_ptr_service = next(
+            (s for s in services if s.name == "test_char_ptr_lambda"), None
+        )
+        assert char_ptr_service is not None, "test_char_ptr_lambda service not found"
+
+        # Execute all four services to test different lambda return types
         client.execute_service(string_service, {"input_string": "STRING_FROM_LAMBDA"})
         client.execute_service(int_service, {"input_number": 42})
         client.execute_service(float_service, {"input_float": 3.14})
+        client.execute_service(
+            char_ptr_service, {"input_number": 123, "input_string": "test_string"}
+        )
 
         # Wait for all service log messages
         # This confirms the lambdas compiled successfully and executed
         try:
             await asyncio.wait_for(
                 asyncio.gather(
-                    string_called_future, int_called_future, float_called_future
+                    string_called_future,
+                    int_called_future,
+                    float_called_future,
+                    char_ptr_called_future,
                 ),
                 timeout=5.0,
             )

